src/Orleans.Core/Core/ClusterClient.cs

```diff
@@ -37,8 +41,26 @@ public ClusterClient(OutsideRuntimeClient runtimeClient, ClientConfiguration con
         {
             this.Configuration = configuration;
             this.runtimeClient = runtimeClient;
+            this.clusterClientLifecycle = new ClusterClientLifecycle(loggerFactory);
+
             //set PropagateActivityId flag from node cofnig
             RequestContext.PropagateActivityId = configuration.PropagateActivityId;
+
+            // register all lifecycle participants
+            IEnumerable<ILifecycleParticipant<IClusterClientLifecycle>> lifecycleParticipants = this.ServiceProvider.GetServices<ILifecycleParticipant<IClusterClientLifecycle>>();
+            foreach (ILifecycleParticipant<IClusterClientLifecycle> participant in lifecycleParticipants)
+            {
+                participant.Participate(clusterClientLifecycle);
+            }
+
+            // register all named lifecycle participants
+            IEnumerable<IKeyedServiceCollection<string, ILifecycleParticipant<IClusterClientLifecycle>>> namedLifecycleParticipantCollections = this.ServiceProvider.GetServices<IKeyedServiceCollection<string, ILifecycleParticipant<IClusterClientLifecycle>>>();
+            foreach (ILifecycleParticipant<IClusterClientLifecycle> participant in namedLifecycleParticipantCollections
+                .SelectMany(c => c.GetServices(this.ServiceProvider)
+                .Select(s => s.GetService(this.ServiceProvider))))
```

Comment:
I think it's valid for a service provider factory to return `null` (you're relying on it in the change to `VersionStoreGrain`), so maybe this should be `s?.GetService(...)` and have a `.Where(s => s != null)` clause added